lib/client.js

```diff
@@ -1,19 +1,108 @@
-var version      = require('version'),
-    Promise      = require('../vendor/ayepromise'),
-    requestCount = 0;
+var PROMISE_TIMEOUT = 5000, // 5 seconds
+    ZAF_EVENT       = /^zaf\./,
+    version         = require('version'),
+    Promise         = window.Promise || require('../vendor/ayepromise'),
+    pendingPromises = {},
+    promiseCount    = 0,
+    requestCount    = 0;
+
+function rawPostMessage(client, msg, forceReady) {
+  if (client.ready || forceReady) {
+    client._source.postMessage(msg, client._origin);
+  } else {
+    client.on('app.registered', rawPostMessage.bind(null, client, msg));
+  }
+}
+
+function wrappedPostMessage(name) {
+  var id = ++promiseCount;
+  var deferred = Promise.defer();
+  var params = Array.prototype.slice.call(arguments, 1);
+
+  // ensure promise is cleaned up when resolved
+  deferred.promise.then(removePromise.bind(null, id), removePromise.bind(null, id));
+
+  // timeout the promise to ensure it will be garbage collected if nobody responds
+  setTimeout(function() {
+    deferred.reject(new Error('Invocation request timeout'));
+  }, PROMISE_TIMEOUT);
+
+  pendingPromises[id] = deferred;
+
+  var msg = JSON.stringify({ id: id, request: name, params: params, appGuid: this._appGuid });
+  rawPostMessage(this, msg);
+
+  return deferred.promise;
+}
+
+function removePromise(id) {
+  delete pendingPromises[id];
+}
+
+function isValidEvent(client, event) {
+  return client._origin === event.origin && client._source === event.source;
+}
+
+function messageHandler(client, event) {
+  if (!isValidEvent(client, event)) { return; }
+
+  var data = event.data;
+
+  if (!data) { return; }
+
+  if (typeof data === 'string') {
+    try {
+      data = JSON.parse(event.data);
+    } catch (e) {
+      return;
+    }
+  }
+
+  console.debug('SDK <<', data);
```

Comment:
Probably not, we could use something like Lotus' minilog, where it only logs when enabled. I wouldn't oppose to removing it for now.